src/devices/Common/System/Device/Analog/AnalogInputPin.cs

```diff
@@ -0,0 +1,179 @@
+        public abstract uint ReadRaw();
+
+        /// <summary>
+        /// Read a raw value and convert it to a voltage
+        /// </summary>
+        /// <returns>Reads a new value from the input and converts it to a voltage. For many boards, the <see cref="VoltageReference"/> needs to be correctly set for this to work.</returns>
+        public virtual ElectricPotential ReadVoltage()
+        {
+            uint raw = ReadRaw();
+            return ConvertToVoltage(raw);
+        }
+
+        /// <summary>
+        /// Converts an input raw value to a voltage
+        /// </summary>
+        protected virtual ElectricPotential ConvertToVoltage(uint rawValue)
+        {
+            if (MinVoltage >= ElectricPotential.Zero)
+            {
+                // The ADC can handle only positive values
+                int maxRawValue = (1 << AdcResolutionBits) - 1;
+                return ((double)rawValue / maxRawValue) * MaxVoltage;
+            }
+            else
+            {
+                // The ADC also handles negative values. This means that the number of bits includes the sign.
+                uint maxRawValue = (uint)((1 << (AdcResolutionBits - 1)) - 1);
+                if (rawValue < maxRawValue)
+                {
+                    return ((double)rawValue / maxRawValue) * MaxVoltage;
+                }
+                else
+                {
+                    // This is a bitmask which has all the bits 1 that are not used in the data.
+                    // We use this to sign-extend our raw value. The mask also includes the sign bit itself,
+                    // but we know that this is already 1
+                    uint topBits = ~maxRawValue;
+                    rawValue |= topBits;
+                    int raw2 = (int)rawValue;
+                    return ((double)raw2 / maxRawValue) * MaxVoltage; // result is negative
+                }
+            }
+        }
+
+        /// <summary>
+        /// Enable event callback when the value of this pin changes.
+        /// </summary>
+        /// <param name="masterController">If an external interrupt handler is required, it can be provided here. Can be null if another interrupt feature is available</param>
+        /// <param name="masterPin">Input pin on the master controller.</param>
+        public abstract void EnableAnalogValueChangedEvent(GpioController? masterController, int masterPin);
+
+        /// <summary>
+        /// Disables the event callback
+        /// </summary>
+        public abstract void DisableAnalogValueChangedEvent();
+
+        /// <summary>
+        /// Fires a value changed event.
+        /// </summary>
+        /// <param name="eventArgs">New value</param>
+        protected void FireValueChanged(ValueChangedEventArgs eventArgs)
+        {
+            ValueChanged?.Invoke(this, eventArgs);
+        }
+
+        /// <summary>
+        /// Close the pin
+        /// </summary>
+        public virtual void Close()
```

Comment:
this shouldn't be public. The only reason we have both Close and Dispose on the Stream or XML is because it was a misdesign - Dispose didn't exist at that time and was added later and Close methods are there for legacy reason. There is no reason to have that in the new code. They semantically mean the same thing.